### DISCUSSION TITLE
Output not as expected from neural network

### DISCUSSION THREAD

**Discussion by finnfi** [bitcraze/discussions#155] [Q&A]
Hi!
I am still struggling to get the gap8 on the crazyflie AI deck to work like I want. So before I take some holiday, Ill post this "issue" here, so maybe someone can help me. I've posted the same thing on the gap_sdk repo from greenwaves. 
 
I have designed a nn in tensorflow and have converted it to a tflite model. 
Then I have used the nntool to quantize and to to create a model for the autotiler. 
I have then written my custom code for the gap 8 and compiled together with the autotiler auto generated code. I had to use the -Os flag to minimize the space the program needs. 
I then run the code on the gap8 on the crazyflie, and it runs (inference takes approximately 10s so it has to be optimized), but the output is trash. 

What I have checked: 
- Compared the quantized vs floating point model with the dump command from the nntool. The quantized output looks ok!
- Checked if the inputs are as they shall before task is sent to cluster. That looks okay as well. 

A complete build log can be found here: [log.txt](https://github.com/GreenWaves-Technologies/gap_sdk/files/9167640/log.txt)
Other files that can be useful: 
[main.zip](https://github.com/GreenWaves-Technologies/gap_sdk/files/9167643/main.zip)
[makefiles.zip](https://github.com/GreenWaves-Technologies/gap_sdk/files/9167645/makefiles.zip)
[depth_estimation_tflite.zip](https://github.com/GreenWaves-Technologies/gap_sdk/files/9167653/depth_estimation_tflite.zip)
[qshow.txt](https://github.com/GreenWaves-Technologies/gap_sdk/files/9167658/qshow.txt)

Appreciate every help. 
Finn

**Comment by knmcguire** [comment]
Hi!

This seems a bit tough to figure out what is going on.. I'm not sure if the -Os flag is doing anything it is not supposed to, so in that sense it sounds like a Greenwave support question.

Have you tried any smaller neural network that fits better to see if you still have problems?

**Reply by finnfi** [reply]
Ye I've run the classification project, which works fine. I may try another network with picture output when Im back from holidays. Havent found one with multiple input yet, so that may be a problem.

**Reply by knmcguire** [reply]
alright! let us know what comes out of it. It would be nice to have more inputs to the NN than only the picture.

**Reply by finnfi** [reply]
My summer job comes to an end, so I will not have so much time to work on this when my studies starts. But I have one little status update on this one: 
Autotiler seems to have some problems with neural networks where layers SHARE WEIGHTS. Our network takes two pictures, and uses the same weights for both encoding parts. When I changed the network to not use any shared weights between the layers, it seems to work. I sadly did not have any time to retrain the network (this bug did take 2-3 weeks to find! :O ), but my hope is that greenwaves fixes this. Issue at Greenwaves: https://github.com/GreenWaves-Technologies/gap_sdk/issues/339

[PS: really like the drone and it has been fun to work with it this summer. Maybe I have some sparetime to explore more. Anyway, greetings from Noway and good luck with the development!]

**Reply by knmcguire** [reply]
HI! Thanks for the update. Perhaps you can quote the one part with the problem and point that as an answer for this thread? Then it will be easier for people to find the issue and that it is related to gap_sdk's issue
